### PR TITLE
Array: refactor `fromIterable` to accept an optional mapping function…

### LINKED ISSUE #2538: Array fromIterable mapfn
### What is the problem this feature would solve?

Without that, people must write `globalThis.Array` or rename Effect import to use Array.from with mapfn.

### What is the feature you are proposing to solve the problem?

Array.from has a second mapfn arg, and because Effect Array.fromIterable uses native Array.from, it should mimic its API imho.

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by gcanti**
… for transformation, closes #2538

## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related

- Closes #2538

### CODE REVIEW

File: packages/effect/src/Array.ts [hunk 1]
```diff
@@ -119,8 +119,11 @@ export const replicate: {
  * @category constructors
  * @since 2.0.0
  */
-export const fromIterable = <A>(collection: Iterable<A>): Array<A> =>
-  Array.isArray(collection) ? collection : Array.from(collection)
+export function fromIterable<A, B>(collection: Iterable<A>, f: (a: A, i: number) => B): Array<B>
+export function fromIterable<A>(collection: Iterable<A>): Array<A>
+export function fromIterable<A>(collection: Iterable<A>, f?: (a: A, i: number) => A): Array<A> {
+  return f ? Array.from(collection, f) : Array.isArray(collection) ? collection : Array.from(collection)
+}
```
Comment (fubhy) [inline packages/effect/src/Array.ts:126]: This is imho a breaking change because it add a new position parameter that can cause problems with tacit usage.

**Review by fubhy — CHANGES_REQUESTED** [review]

File: packages/effect/src/Array.ts (hunk 1 above)
Comment (gcanti) [reply packages/effect/src/Array.ts:126]: Yes, I haven't changed the test to show that it's a breaking change, I think we should close this and tag the corresponding issue with a "next-major" label.

Comment (fubhy) [reply packages/effect/src/Array.ts:126]: Agreed

Comment (mikearnaldi) [reply packages/effect/src/Array.ts:126]: I don't think we generally support tacit usage

**Pull request closed without merging**